README.md
# Starburst Landing Page

A small web server that serves a landing page for a collection of devices.

Supports pinging devices for availability, and sending wake-on-lan packets.

Useful for home networks with a collection of devices that can be accessed
remotely and are not always on, like desktops and laptops.

## Requirements

- [Deno](https://deno.land/)
- [Node.js](https://nodejs.org/en/)
- [Yarn](https://yarnpkg.com/)

## Getting Started

### Docker

> If you want to support wake-on-lan, run the `/scripts/hash_password.ts` script
> to generate a password hash and salt.
>
> ```sh
> docker run --rm -it -v ghcr.io/lixquid/starburst-landingpage:latest deno run --allow-env /scripts/hash_password.ts
> ```

1. Copy the `backend/config.example.json` file to `config.json`, and update the
   values as needed.
2. Run the server.

   ```sh
   docker run -v "$(pwd)/config.json:/app/config.json" -p 8080:8080 ghcr.io/lixquid/starburst-landingpage:latest
   ```

### From Source

> If you want to support wake-on-lan, run the `scripts/hash_password.ts` script
> to generate a password hash and salt.
>
> ```sh
> deno run --allow-env scripts/hash_password.ts
> ```

1. Run the `scripts/package_frontend.ts` script to build and package the
   frontend into the correct locations in the backend.

   ```sh
   deno run --allow-read --allow-write --allow-run scripts/package_frontend.ts
   ```
2. Copy the `backend/config.example.json` file to `backend/config.json`, and
   update the values as needed.
3. Run the server.

   ```sh
   cd backend
   deno run --allow-net --allow-read --unstable main.ts
   ```

For more information, all scripts and the server support the `--help` flag.

<details>
<summary><h2>Developer Documentation</h2></summary>

### Preparing a new release

1. Create an entry in [CHANGELOG.md](CHANGELOG.md).
2. Update the README.md file as needed.
3. Commit with `Version x.y.z` as the commit message.
4. Tag the commit with `vX.Y.Z`.

</details>


CHANGELOG.md
# Version 0.1.0

- Support for pinging devices for availability.
- Support for sending wake-on-lan packets.


frontend/vite.config.ts
import { Eta } from "eta";
import { defineConfig } from "vite";

const demoData = {
    name: "Demo Landing",
    servers: [
        {
            name: "Inactive",
            active: false,
            controllable: false,
        },
        {
            name: "Active",
            active: true,
            controllable: false,
        },
        {
            name: "Controllable",
            active: true,
            controllable: true,
        },
    ],
};

export default defineConfig({
    plugins: [
        // This plugin renders the eta tags in the index.html file with
        // some demo data in development mode.
        {
            name: "render-eta-in-dev",
            apply: "serve",
            transformIndexHtml: {
                enforce: "pre",
                async handler(html) {
                    return new Eta({
                        tags: ["{{", "}}"],
                    }).renderStringAsync(html, demoData);
                },
            },
        },
    ],
});


scripts/hash_password.ts
// Hashes a password for use with the Starburst Landing Page server.
// Run with --help for more information.

import * as flags from "https://deno.land/std@0.193.0/flags/mod.ts";
import { hash } from "https://deno.land/x/argontwo@0.1.1/mod.ts";

const { password, salt, memory, iterations, parallelism, json } =
    await parseArgs();

const encoder = new TextEncoder();

const hashRaw = hash(
    encoder.encode(password),
    encoder.encode(salt),
    {
        m: memory,
        t: iterations,
        p: parallelism,
    },
);
const hashOutput = Array.from(hashRaw).map((b) =>
    b.toString(16).padStart(2, "0")
).join("");

if (json) {
    console.log(JSON.stringify({
        hash: hashOutput,
        salt,
        memory,
        iterations,
        parallelism,
    }));
} else {
    console.log(`Hash: ${hashOutput}`);
    console.log(`Salt: ${salt}`);
    console.log();
    console.log(`Memory: ${memory}`);
    console.log(`Iterations: ${iterations}`);
    console.log(`Parallelism: ${parallelism}`);
}

//#region Utility Functions
const helpMessage = `hash_password.ts [options] [password|-]

Hashes a password for use with the Starburst Landing Page server.

NOTE: Passwords will have leading and trailing whitespace removed.

Options with a STARBURST_ word in all caps can be set as environment variables.

Options:
    password                The password to hash.
    STARBURST_PASSWORD      If "-", the password will be read from stdin.
                            Otherwise, the password will be prompted for.

    -s, --salt              The salt to use for hashing.
    STARBURST_SALT          If not set, a random salt will be generated.

    -j, --json              Output the result as JSON.
    STARBURST_JSON

    -h, --help              Show this help message.

Advanced Options:
    --memory                The amount of memory to use for hashing, in KiB.
    STARBURST_MEMORY        Defaults to 4096.

    --iterations            The number of iterations to use for hashing.
    STARBURST_ITERATIONS    Defaults to 3.

    --parallelism           The number of threads to use for hashing.
    STARBURST_PARALLELISM   Defaults to 1.
`;

async function parseArgs() {
    const args = flags.parse(Deno.args, {
        string: ["salt", "memory", "iterations", "parallelism"],
        boolean: ["help", "json"],
        alias: {
            help: "h",
            salt: "s",
            json: "j",
        },
    });

    if (args.help) {
        console.log(helpMessage);
        Deno.exit(-1);
    }

    const password = await getPassword(args._[0]);

    const salt = args.salt ?? Deno.env.get("STARBURST_SALT") ?? generateSalt();

    const memory = parseInt(
        args.memory ?? Deno.env.get("STARBURST_MEMORY") ?? "4096",
    );
    if (isNaN(memory)) {
        console.error("Invalid memory value");
        Deno.exit(1);
    }

    const iterations = parseInt(
        args.iterations ?? Deno.env.get("STARBURST_ITERATIONS") ?? "3",
    );
    if (isNaN(iterations)) {
        console.error("Invalid iterations value");
        Deno.exit(1);
    }

    const parallelism = parseInt(
        args.parallelism ?? Deno.env.get("STARBURST_PARALLELISM") ?? "1",
    );
    if (isNaN(parallelism)) {
        console.error("Invalid parallelism value");
        Deno.exit(1);
    }

    const json = args.json ?? !!Deno.env.get("STARBURST_JSON") ?? false;

    return { password, salt, memory, iterations, parallelism, json };
}

async function getPassword(pass: string | number) {
    switch (pass) {
        case "-": {
            // Read from stdin
            const buf = new Uint8Array(1024);
            let input = "";
            while (true) {
                const n = await Deno.stdin.read(buf);
                if (n === null) {
                    break;
                }
                input += new TextDecoder().decode(buf.subarray(0, n));
            }
            return input.trim();
        }
        case undefined: {
            // Check environment variable
            const password = Deno.env.get("STARBURST_PASSWORD");
            if (password !== undefined) {
                return password.trim();
            }

            // Prompt the user
            const response = prompt("Enter password:");
            if (response !== null) {
                return response.trim();
            }

            console.error("No password provided");
            Deno.exit(1);
            break;
        }
        default: {
            // Use the argument
            return String(pass).trim();
        }
    }
}

function generateSalt() {
    return Array.from(crypto.getRandomValues(new Uint8Array(16)))
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
}
//#endregion


scripts/package_frontend.ts
// Packages the frontend into the correct location for the backend to serve it.
// Run with --help for more information.

import * as flags from "https://deno.land/std@0.193.0/flags/mod.ts";
import * as fs from "https://deno.land/std@0.193.0/fs/mod.ts";
import * as path from "https://deno.land/std@0.193.0/path/mod.ts";

parseArgs();

const root = path.dirname(path.dirname(path.fromFileUrl(import.meta.url)));
log("Project Root:", root);

// Check npx is available
log("Checking npx is available...");
await run(["npx", "--version"]);

// Ensure the frontend dependencies are installed
log("Installing frontend dependencies...");
await run(["npx", "yarn", "install"]);

// Clear the "dist" directory
log("Clearing frontend/dist directory...");
await fs.emptyDir(path.join(root, "frontend", "dist"));

// Run the "yarn build" command
log("Building frontend...");
await run(["npx", "yarn", "build"]);

// Remove the backend/static directory
log("Removing backend/static directory...");
await Deno.remove(path.join(root, "backend", "static"), { recursive: true });

// Copy the frontend files to the backend
log("Copying frontend files to backend...");
await fs.copy(
    path.join(root, "frontend", "dist"),
    path.join(root, "backend", "static"),
    { overwrite: true },
);

// Ensure the backend/templates directory exists
log("Ensuring backend/templates directory exists...");
await fs.ensureDir(path.join(root, "backend", "templates"));

// Empty the backend/templates directory
log("Emptying backend/templates directory...");
await fs.emptyDir(path.join(root, "backend", "templates"));

// Move the HTML files to the templates directory
log("Moving HTML files to templates directory...");
for await (
    const entry of Deno.readDir(
        path.join(root, "backend", "static"),
    )
) {
    if (entry.isFile && entry.name.endsWith(".html")) {
        await Deno.rename(
            path.join(root, "backend", "static", entry.name),
            path.join(
                root,
                "backend",
                "templates",
                entry.name.replace(".html", ".eta"),
            ),
        );
    }
}

log("Done!");

//#region Utility Functions
const helpMessage = `package_frontend.ts [options]

Packages the frontend into the correct location for the backend to serve it.

Options:
    -h, --help              Show this help message.
`;

function parseArgs() {
    const args = flags.parse(Deno.args, {
        boolean: [
            "help",
        ],
        alias: {
            "help": ["h"],
        },
    });

    if (args.help) {
        console.log(helpMessage);
        Deno.exit(-1);
    }
}

function log(...args: unknown[]) {
    console.log("►", ...args);
}

async function run(cmd: string[]) {
    // If windows, use cmd.exe
    const isWindows = Deno.build.os === "windows";
    if (isWindows) {
        cmd = ["cmd.exe", "/c", ...cmd];
    }

    const [command, ...args] = cmd;
    const process = new Deno.Command(command, {
        args,
        cwd: path.join(root, "frontend"),
    }).spawn();
    const status = await process.status;

    if (!status.success) {
        console.error(`Failed to run command: ${cmd.join(" ")}`);
        Deno.exit(1);
    }
}
//#endregion


backend/main.ts
// A server to see the status of devices and send wake-on-lan packets.
// Run with --help to see options.

import * as flags from "https://deno.land/std@0.193.0/flags/mod.ts";
import * as path from "https://deno.land/std@0.193.0/path/mod.ts";
import { hash } from "https://deno.land/x/argontwo@0.1.1/mod.ts";
import { Eta } from "https://deno.land/x/eta@v3.0.3/src/index.ts";
import * as oak from "https://deno.land/x/oak@v12.5.0/mod.ts";
import * as wol from "https://deno.land/x/wol@v1.0.1/mod.ts";
// @deno-types="npm:@types/ping"
import ping from "npm:ping@0.4.4";

const { configPath, port, verbose } = await parseArgs();
const config = await readConfig(configPath);
if (verbose) printConfig(port, config);

const __dirname = path.dirname(path.fromFileUrl(import.meta.url));
const eta = new Eta({
    views: path.join(__dirname, "templates"),
    tags: ["{{", "}}"],
});

const app = new oak.Application();
const router = new oak.Router();

//// Logging ////
if (verbose) {
    app.use(async (ctx, next) => {
        await next();
        console.log(
            `${ctx.request.method} ${ctx.request.url} - ${ctx.response.status}`,
        );
    });
}

//// WOL Requests ////
router.get("/wol/:index", async (ctx) => {
    const index = parseInt(ctx.params.index);
    if (isNaN(index) || index < 0 || index >= config.servers.length) {
        ctx.response.status = 400;
        ctx.response.body = "Invalid index";
        return;
    }

    // Check authorization
    const auth = ctx.request.headers.get("Authorization");
    if (!matchesPassword(auth)) {
        ctx.response.status = 401;
        ctx.response.headers.set("WWW-Authenticate", "Basic");
        ctx.response.body = "Unauthorized";
        return;
    }

    // Check if server supports WOL
    const mac = config.servers[index].mac;
    if (mac === undefined) {
        ctx.response.status = 400;
        ctx.response.body = "Server does not support WOL";
        return;
    }

    // Send WOL packet
    try {
        await wol.wake(mac);
        if (verbose) console.log(`Sent WOL packet to ${mac}`);
    } catch (ex: unknown) {
        ctx.response.status = 500;
        ctx.response.body = "Failed to send WOL packet";
        console.error(`Failed to send WOL packet to ${mac}: ${ex}`);
        return;
    }

    ctx.response.body = await renderIndex();
});

//// Index Page ////
router.get("/", async (ctx) => {
    ctx.response.body = await renderIndex();
});

app.use(router.routes());

//// Static Files ////
app.use(async (ctx, next) => {
    try {
        await ctx.send({
            root: path.join(__dirname, "static"),
        });
    } catch {
        await next();
    }
});

//// Start Server ////
app.addEventListener("listen", () => {
    console.log(`Listening on port ${port}`);
});
await app.listen({ port });

//#region Utility Methods
const helpMessage = `main.ts [options]

A server to see the status of devices and send wake-on-lan packets.

Options with a STARBURST_ word in all caps can be set as environment variables.

Options:
    -h, --help                Show this help message.

    -v, --verbose             Enable verbose logging.
    STARBURST_VERBOSE

    -c, --config              The path to the config file.
    STARBURST_CONFIG          Defaults to "config.json".

    -p, --port                The port to listen on.
    STARBURST_PORT            Defaults to 8080.
`;

/** Parses the command line arguments. */
async function parseArgs() {
    const args = flags.parse(Deno.args, {
        string: [
            "config",
            "port",
        ],
        boolean: [
            "help",
            "verbose",
        ],
        alias: {
            h: "help",
            c: "config",
            p: "port",
            v: "verbose",
        },
    });

    if (args.help) {
        console.log(helpMessage);
        Deno.exit(-1);
    }

    const configPath = args.config ?? Deno.env.get("STARBURST_CONFIG") ??
        "config.json";
    if (await Deno.stat(configPath).catch(() => null) === null) {
        console.error(`Config file "${configPath}" does not exist`);
        Deno.exit(-1);
    }

    const port = parseInt(
        args.port ?? Deno.env.get("STARBURST_PORT") ?? "8080",
    );
    if (isNaN(port)) {
        console.error("Invalid port value");
        Deno.exit(-1);
    }
    if (port < 0 || port > 65535) {
        console.error("Port must be between 0 and 65535");
        Deno.exit(-1);
    }

    return {
        configPath,
        port,
        verbose: args.verbose ?? !!Deno.env.get("STARBURST_VERBOSE"),
    };
}

/** Reads and validates the config file and returns the config object. */
async function readConfig(path: string) {
    const config = JSON.parse(await Deno.readTextFile(path));
    if (typeof config !== "object" || config === null) {
        console.error("Invalid config file");
        Deno.exit(-1);
    }

    const name = config.name as unknown ?? undefined;
    if (name !== undefined && typeof name !== "string") {
        console.error(`Invalid config: "name" must be a string or unset`);
        Deno.exit(-1);
    }

    const passwordHash = config.password?.hash as unknown ?? undefined;
    if (passwordHash !== undefined && typeof passwordHash !== "string") {
        console.error(
            `Invalid config: "password.hash" must be a string or unset`,
        );
        Deno.exit(-1);
    }
    if (passwordHash !== undefined && !/^[0-9a-f]{64}$/i.test(passwordHash)) {
        console.error(
            `Invalid config: "password.hash" must be a 64-character hex string`,
        );
        Deno.exit(-1);
    }

    const passwordSalt = config.password?.salt as unknown ?? undefined;
    if (passwordSalt !== undefined && typeof passwordSalt !== "string") {
        console.error(
            `Invalid config: "password.salt" must be a string or unset`,
        );
        Deno.exit(-1);
    }

    const passwordMemory = config.password?.advanced?.memory as unknown ?? 4096;
    if (typeof passwordMemory !== "number" || isNaN(passwordMemory)) {
        console.error(
            `Invalid config: "password.advanced.memory" must be a number or unset`,
        );
        Deno.exit(-1);
    }
    if (passwordMemory < 0) {
        console.error(
            `Invalid config: "password.advanced.memory" must be positive`,
        );
        Deno.exit(-1);
    }

    const passwordIterations =
        config.password?.advanced?.iterations as unknown ?? 3;
    if (typeof passwordIterations !== "number" || isNaN(passwordIterations)) {
        console.error(
            `Invalid config: "password.advanced.iterations" must be a number or unset`,
        );
        Deno.exit(-1);
    }
    if (passwordIterations < 0) {
        console.error(
            `Invalid config: "password.advanced.iterations" must be positive`,
        );
        Deno.exit(-1);
    }

    const passwordParallelism =
        config.password?.advanced?.parallelism as unknown ?? 1;
    if (typeof passwordParallelism !== "number" || isNaN(passwordParallelism)) {
        console.error(
            `Invalid config: "password.advanced.parallelism" must be a number or unset`,
        );
        Deno.exit(-1);
    }
    if (passwordParallelism < 0) {
        console.error(
            `Invalid config: "password.advanced.parallelism" must be positive`,
        );
        Deno.exit(-1);
    }

    const serversRaw = config.servers as unknown;
    if (!Array.isArray(serversRaw)) {
        console.error(`Invalid config: "servers" must be an array`);
        Deno.exit(-1);
    }

    const servers = serversRaw.map((v, i) => {
        if (typeof v !== "object" || v === null) {
            console.error(`Invalid config: "servers[${i}]" must be an object`);
            Deno.exit(-1);
        }

        const name = v.name as unknown;
        if (typeof name !== "string") {
            console.error(
                `Invalid config: "servers[${i}].name" must be a string`,
            );
            Deno.exit(-1);
        }

        const hostname = v.hostname as unknown;
        if (typeof hostname !== "string") {
            console.error(
                `Invalid config: "servers[${i}].hostname" must be a string`,
            );
            Deno.exit(-1);
        }

        const mac = v.mac as unknown ?? undefined;
        if (mac !== undefined && typeof mac !== "string") {
            console.error(
                `Invalid config: "servers[${i}].mac" must be a string or unset`,
            );
            Deno.exit(-1);
        }

        return { name, hostname, mac };
    });

    return {
        name,
        passwordHash,
        passwordSalt,
        passwordMemory,
        passwordIterations,
        passwordParallelism,
        servers,
    };
}

/** Checks if the given auth header matches the password in the config. */
function matchesPassword(authHeader: string | null): boolean {
    if (!config.passwordHash || !config.passwordSalt) {
        return false;
    }
    if (authHeader === null) {
        return false;
    }
    if (!authHeader.startsWith("Basic ")) {
        return false;
    }
    authHeader = authHeader.slice("Basic ".length);
    const [, password] = atob(authHeader).split(":", 2);

    const encoder = new TextEncoder();
    const hashRaw = hash(
        encoder.encode(password),
        encoder.encode(config.passwordSalt),
        {
            m: config.passwordMemory,
            t: config.passwordIterations,
            p: config.passwordParallelism,
        },
    );
    const hex = Array.from(hashRaw)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");

    return hex === config.passwordHash;
}

/** Prints the config to stdout. */
function printConfig(
    port: number,
    config: Awaited<ReturnType<typeof readConfig>>,
) {
    console.log(`Name: ${config.name ?? "(unset)"}`);
    console.log(`Port: ${port}`);
    console.log(`Password Hash: ${config.passwordHash ?? "(unset)"}`);
    console.log(`Password Salt: ${config.passwordSalt ?? "(unset)"}`);
    console.log(`    Password Memory: ${config.passwordMemory}`);
    console.log(`    Password Iterations: ${config.passwordIterations}`);
    console.log(`    Password Parallelism: ${config.passwordParallelism}`);
    console.log(`Servers:`);
    for (const [i, server] of config.servers.entries()) {
        console.log(`    ${i + 1}:`);
        console.log(`        Name: ${server.name}`);
        console.log(`        Hostname: ${server.hostname}`);
        console.log(`        MAC: ${server.mac ?? "(unset)"}`);
    }
}

/** Renders the index eta template. */
async function renderIndex() {
    const pings = await Promise.all(config.servers.map((server) => {
        return ping.promise.probe(server.hostname, {
            timeout: 1,
            min_reply: 1,
        }) as Promise<{ alive: boolean }>;
    }));
    return await eta.renderAsync("index", {
        name: config.name ?? "",
        servers: config.servers.map((server, i) => ({
            name: server.name,
            active: pings[i].alive,
            controllable: server.mac !== undefined,
        })),
    });
}
//#endregion
